File: Frameworks/Swift4.0.3/SampleCode/objcdemo/objcdemo/ViewController.m
//
//  ViewController.m
//  objcdemo
//
//  Created by gabriel izebhigie on 15/01/2018.
//  Copyright © 2018 Interswitch. All rights reserved.
//

#import "ViewController.h"
@import PaymentSDK;

@interface ViewController ()

@property (strong, nonatomic) UITextField *customerId;
@property (strong, nonatomic) UITextField *amount;
@property (strong, nonatomic) UITextField *pan;
@property (strong, nonatomic) UITextField *expiry;
@property (strong, nonatomic) UITextField *pin;
@property (strong, nonatomic) UITextField *cvv;
@property (strong, nonatomic) UIActivityIndicatorView *activityIndicator;
@property (strong, nonatomic) UIButton *payNow;

@end

@implementation ViewController


NSString *myclientId  = @"IKIA7B379B0114CA57FAF8E19F5CC5063ED2220057EF";
NSString *myclientSecret  = @"MiunSQ5SN219UCVP1Lt2raPfwK9lMoiV/PdBX5v/R4=";

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    [Passport overrideApiBase:@"https://qa.interswitchng.com/passport"];
    [Payment overrideApiBase:@"https://qa.interswitchng.com/passport"];
    
      self.customerId = [[UITextField alloc] initWithFrame:CGRectMake(20, 100, 200, 30)];
      self.amount = [[UITextField alloc] initWithFrame:CGRectMake(20, 140, 200, 30)];
      self.pan = [[UITextField alloc] initWithFrame:CGRectMake(20, 180, 200, 30)];
      self.expiry = [[UITextField alloc] initWithFrame:CGRectMake(20, 220, 200, 30)];
      self.cvv = [[UITextField alloc] initWithFrame:CGRectMake(20, 260, 200, 30)];
      self.pin = [[UITextField alloc] initWithFrame:CGRectMake(20, 300, 200, 30)];
    
      self.activityIndicator = [[UIActivityIndicatorView alloc] initWithFrame:CGRectMake(0,0,40,40)];
    
    self.view.backgroundColor = UIColor.whiteColor;
    
    self.customerId.placeholder = @"Customer ID";
    self.customerId.borderStyle = UITextBorderStyleLine;
    self.customerId.text = @"1407002510";
    [self.view addSubview:self.customerId];
    
    self.amount.placeholder = @"Amount";
    self.amount.borderStyle = UITextBorderStyleLine;
    self.amount.text = @"1";
    [self.view addSubview:self.amount];
    
    self.pan.placeholder = @"Card No";
    self.pan.borderStyle = UITextBorderStyleLine;
    self.pan.keyboardType = UIKeyboardTypeNumberPad;
    self.pan.text = @"5060990580000217499";
    [self.view addSubview:self.pan];
    
    self.pin.placeholder = @"Card PIN";
    self.pin.borderStyle = UITextBorderStyleLine;
    self.pin.keyboardType = UIKeyboardTypeNumberPad;
    self.pin.text = @"1111";
    [self.view addSubview:self.pin];
    
    self.expiry.placeholder = @"Expiry";
    self.expiry.borderStyle = UITextBorderStyleLine;
    self.expiry.keyboardType = UIKeyboardTypeNumberPad;
    self.expiry.text = @"2004";
    [self.view addSubview:self.expiry];
    
    self.cvv.placeholder = @"CVV";
    self.cvv.borderStyle = UITextBorderStyleLine;
    self.cvv.keyboardType = UIKeyboardTypeNumberPad;
    self.cvv.text = @"111";
    [self.view addSubview:self.cvv];
    
    self.activityIndicator.activityIndicatorViewStyle = UIActivityIndicatorViewStyleGray;
    self.activityIndicator.center = self.view.center;
    [self.view addSubview:self.activityIndicator];
    [self.view bringSubviewToFront:self.activityIndicator];
    
    //UIButton *payNow = [[UIButton alloc] initWithFrame:CGRectMake(20, 340, 40, 40)];
    self.payNow = [UIButton buttonWithType:UIButtonTypeCustom];
    self.payNow.frame = CGRectMake(20, 340, 40, 40);
    //[payNow setTitle:@"Pay" forState:UIControlStateSelected];
    self.payNow.titleLabel.font = [UIFont fontWithName:@"Helvetica-Bold" size:22];
    self.payNow.backgroundColor = UIColor.blackColor;
    self.payNow.titleLabel.textColor = UIColor.whiteColor;
    self.payNow.titleLabel.text = @"Pay";
    [self.view addSubview:self.payNow];
    [self.payNow addTarget:self action:@selector(onPayButtonClicked) forControlEvents:UIControlEventTouchUpInside];
    
    self.modalPresentationStyle = UIModalPresentationPopover;
    
}


-(void)onPayButtonClicked {
    if (!self.customerId.hasText) {
        [self showError:@"Customer ID is required"];
    } else if (!self.amount.hasText) {
        [self showError:@"Amount is required"];
    }else if (!self.pan.hasText) {
        [self showError:@"Card No is required"];
    }else if (!self.cvv.hasText) {
        [self showError:@"CVV is required"];
    }else if (!self.pin.hasText) {
        [self showError:@"Card PIN is required"];
    }else if (!self.expiry.hasText) {
        [self showError:@"Expiry is required"];
    }else {
         //[self showSuccess:@"Good Job"];
        
//        PaymentSDK *sdk = [[PaymentSDK alloc] initWithClientId:myclientId clientSecret:myclientSecret];
//
//        PurchaseRequest *purchase = [[PurchaseRequest alloc] initWithCustomerId:self.customerId.text amount:self.amount.text pan:self.pan.text pin:self.pin.text expiryDate:self.expiry.text cvv2:self.cvv.text transactionRef:[Payment randomStringWithLength: 12] currency:@"NGN" requestorId:@"12345678901"];
//
//
//        [sdk purchase:purchase completionHandler:^(PurchaseResponse * _Nullable purchaseResponse, NSError * _Nullable error) {
//
//            if(error != nil) {
//                NSString *errMsg = error.localizedDescription;
//
//                NSLog(@"Error: %@", errMsg);
//            } else if(purchaseResponse == nil) {
//                NSString *errMsg = error.localizedFailureReason;
//
//                NSLog(@"Failure: %@", errMsg);
//            } else if(purchaseResponse.responseCode == nil || [purchaseResponse.responseCode length] == 0) {
//                // OTP not required, payment successful.
//                // The response object contains fields transactionIdentifier, message,
//                // amount, token, tokenExpiryDate, panLast4Digits, otpTransactionIdentifier,
//                // transactionRef and cardType.
//                // Save the token, tokenExpiryDate, cardType and panLast4Digits in order to pay with the token in the future.
//            } else {
//                //NSString *responseCode = validateCardResponse.responseCode;
//                // At this point further authorization is required depending on the value of responseCode
//                // Please see below: Authorize PayWithCard
//            }
//        }];
        
        
//        PayWithCard *payWithCard = [[PayWithCard alloc] initWithClientId:myclientId clientSecret:myclientSecret customerId:self.customerId.text description:@"Payment" amount:self.amount.text currency:@"NGN"];
//
//        UIViewController *viewCtrl =[payWithCard start:^(PurchaseResponse * _Nullable purchaseResponse, NSError * _Nullable error) {
//            [self showError:error.localizedDescription];
//        }];
//
        WalletSDK *walletSdk = [[WalletSDK alloc] initWithClientId:myclientId clientSecret:myclientSecret];
        
        [walletSdk getPaymentMethods:^(WalletResponse *walletResponse, NSError *error) {
            if(error != nil) {
                NSString *errMsg = error.localizedDescription;
                
                NSLog(@"error getting payment methods ... %@", errMsg);
            } else if(walletResponse == nil) {
                NSString *errMsg = error.localizedFailureReason;
                
                NSLog(@"Failure: %@", errMsg);
            } else {
                if ([walletResponse.paymentMethods count] > 0) {
                    for (int i = 0; i < [walletResponse.paymentMethods count]; i++) {
                        PaymentMethod *aPaymentMethod = [walletResponse.paymentMethods objectAtIndex:i];
                        
                        NSLog(@"Payment Method card product: %@", aPaymentMethod.cardProduct);
                    }
                }
            }
        }];
        
    }
    
}

-(void)showError:(NSString *) message {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error" message:message preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *action = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
    [alert addAction:action];
    [self presentViewController:alert animated:YES completion:nil ];
}

-(void)showSuccess:(NSString *) message {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Success" message:message preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *action = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
    [alert addAction:action];
    [self presentViewController:alert animated:YES completion:nil ];
}

-(void)handleOTP:(NSString *) message {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"OTP2" message:message preferredStyle:UIAlertControllerStyleAlert];
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = @"Enter OTP";
    }];
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleDefault handler:nil];
    [alert addAction:cancel];
    
    UIAlertAction *ok = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        if(alert.textFields.firstObject.hasText != nil){
            NSString *otp = alert.textFields.firstObject.text;
            //AuthorizeOtpRequest *otpReq =
        }
    }];
    [alert addAction:ok];
    
    [self presentViewController:alert animated:YES completion:nil ];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: Frameworks/SampleCode/SdkTest/SdkTest/ViewController.swift
//
//  ViewController.swift
//  SdkTest
//
//  Created by gabriel izebhigie on 11/01/2018.
//  Copyright © 2018 Interswitch. All rights reserved.
//

import UIKit
import PaymentSDK

class ViewController: UIViewController {
    
    var customerId: UITextField
    var amount: UITextField
    var pan: UITextField
    var pin: UITextField
    var cvv: UITextField
    var expiry: UITextField
    var activityIndicator: UIActivityIndicatorView
    var otpTransactionIdentifier: String?
    var transactionIdentifier: String?
    let sdk = PaymentSDK(clientId: "IKIA7B379B0114CA57FAF8E19F5CC5063ED2220057EF", clientSecret: "MiunSQ5S/N219UCVP1Lt2raPfwK9lMoiV/PdBX5v/R4=")
    
    required init?(coder aDecoder: NSCoder) {
        customerId = UITextField(frame: CGRect(x: 20, y: 100, width: 200, height: 30))
        amount = UITextField(frame: CGRect(x: 20, y: 140, width: 200, height: 30))
        pan = UITextField(frame: CGRect(x: 20, y: 180, width: 200, height: 30))
        expiry = UITextField(frame: CGRect(x: 20, y: 220, width: 200, height: 30))
        cvv = UITextField(frame: CGRect(x: 20, y: 260, width: 200, height: 30))
        pin = UITextField(frame: CGRect(x: 20, y: 300, width: 200, height: 30))
        activityIndicator = UIActivityIndicatorView(frame: CGRect(x: 0, y: 0, width: 40, height: 40))
        super.init(coder: aDecoder)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        // Do any additional setup after loading the view.
        Passport.overrideApiBase("https://qa.interswitchng.com/passport")
        Payment.overrideApiBase("https://qa.interswitchng.com")
        view.backgroundColor = UIColor.white
        customerId.placeholder = "Customer ID"
        customerId.borderStyle = UITextBorderStyle.line
        customerId.text = "1407002510"
        view.addSubview(customerId)
        
        amount.placeholder = "Amount"
        amount.borderStyle = UITextBorderStyle.line
        amount.keyboardType = .decimalPad
        amount.text = "1"
        view.addSubview(amount)
        
        pan.placeholder = "Card No"
        pan.borderStyle = UITextBorderStyle.line
        pan.keyboardType = .numberPad
        pan.text = "5060990580000217499"
        view.addSubview(pan)
        
        pin.placeholder = "Card PIN"
        pin.borderStyle = UITextBorderStyle.line
        pin.keyboardType = .numberPad
        pin.text = "1111"
        view.addSubview(pin)
        
        expiry.placeholder = "Expiry"
        expiry.borderStyle = UITextBorderStyle.line
        expiry.keyboardType = .numberPad
        expiry.text = "2004"
        view.addSubview(expiry)
        
        cvv.placeholder = "CVV"
        cvv.borderStyle = UITextBorderStyle.line
        cvv.keyboardType = .numberPad
        cvv.text = "111"
        view.addSubview(cvv)
        activityIndicator.activityIndicatorViewStyle = UIActivityIndicatorViewStyle.gray
        activityIndicator.center = view.center
        view.addSubview(activityIndicator)
        activityIndicator.bringSubview(toFront: view)
        
        
        let payNow = UIButton(type: UIButtonType.roundedRect)
        payNow.frame = CGRect(x: 20, y: 340, width: 40, height: 40)
        payNow.setTitle("Pay", for: UIControlState())
        payNow.titleLabel?.font = UIFont.boldSystemFont(ofSize: 22)
        payNow.addTarget(self, action: #selector(ViewController.pay), for: UIControlEvents.touchUpInside)
        view.addSubview(payNow)
        modalPresentationStyle = .popover
    }
    
    @objc func pay(){
        if !customerId.hasText {
            showError("Customer ID is required")
        } else if !amount.hasText{
            showError("Amount is required")
        }else if !pan.hasText{
            showError("Card No is required")
        }else if !cvv.hasText{
            showError("CVV is required")
        }else if !pin.hasText{
            showError("Card PIN is required")
        }else if !expiry.hasText{
            showError("Expiry is required")
        }else{
            let request = PurchaseRequest(customerId: customerId.text, amount: amount.text!, pan: pan.text!, pin: pin.text!, expiryDate: expiry.text!, cvv2: cvv.text!, transactionRef: Payment.randomStringWithLength(12), requestorId: "12345678901")
            
            
            UIApplication.shared.isNetworkActivityIndicatorVisible = true
            activityIndicator.startAnimating()
            sdk.purchase(request, completionHandler:{(purchaseResponse: PurchaseResponse?, error: Error?) in
                self.otpTransactionIdentifier = "1234567"
                guard error == nil else {
                    UIApplication.shared.isNetworkActivityIndicatorVisible = false
                    self.activityIndicator.stopAnimating()
                    self.showError((error?.localizedDescription)!)
                    return
                }
                
                guard let response = purchaseResponse else {
                    UIApplication.shared.isNetworkActivityIndicatorVisible = false
                    self.activityIndicator.stopAnimating()
                    self.showError((error?.localizedDescription)!)
                    return
                }
                guard let otpTransactionIdentifier = response.otpTransactionIdentifier else {
                    UIApplication.shared.isNetworkActivityIndicatorVisible = false
                    self.activityIndicator.stopAnimating()
                    self.showSuccess("Ref: " + response.transactionIdentifier)
                    return
                }
                self.otpTransactionIdentifier = otpTransactionIdentifier
                self.transactionIdentifier = response.transactionIdentifier
                self.handleOTP(response.message)
            } )
            
        }
    }
    func showError(_ message: String){
        let alert = UIAlertController(
            title: "Error",
            message: message,
            preferredStyle: UIAlertControllerStyle.alert
        )
        let action = UIAlertAction(title: "OK", style: UIAlertActionStyle.default, handler: nil)
        alert.addAction(action)
        present(alert, animated: true, completion: nil)
    }
    
    func showSuccess(_ message: String){
        let alert = UIAlertController(
            title: "Success",
            message: message,
            preferredStyle: UIAlertControllerStyle.alert
        )
        let action = UIAlertAction(title: "OK", style: UIAlertActionStyle.default, handler: nil)
        alert.addAction(action)
        present(alert, animated: true, completion: nil)
    }
    
    func handleOTP(_ message: String){
        let alert = UIAlertController(
            title: "OTP2",
            message: message,
            preferredStyle: UIAlertControllerStyle.alert
        )
        alert.addTextField { (textField) -> Void in
            textField.placeholder = "Enter OTP"
        }
        let cancel = UIAlertAction(title: "Cancel", style: UIAlertActionStyle.default, handler: nil)
        alert.addAction(cancel)
        let ok = UIAlertAction(title: "OK", style: UIAlertActionStyle.default, handler: {(action) -> Void in
            if ((alert.textFields?.first?.hasText) != nil){
                let otp = alert.textFields?.first?.text
                let otpReq = AuthorizeOtpRequest(otpTransactionIdentifier: self.otpTransactionIdentifier!, otp: otp!, transactionRef: Payment.randomStringWithLength(12))
                UIApplication.shared.isNetworkActivityIndicatorVisible = true
                self.activityIndicator.startAnimating()
                self.sdk.authorizeOtp(otpReq, completionHandler: {(authorizeOtpResponse: AuthorizeOtpResponse?, error: Error?) in
                    guard error == nil else {
                        UIApplication.shared.isNetworkActivityIndicatorVisible = false
                        self.activityIndicator.stopAnimating()
                        self.showError((error?.localizedDescription)!)
                        return
                    }
                    
                    guard authorizeOtpResponse != nil else {
                        UIApplication.shared.isNetworkActivityIndicatorVisible = false
                        self.activityIndicator.stopAnimating()
                        self.showError((error?.localizedDescription)!)
                        return
                    }
                    UIApplication.shared.isNetworkActivityIndicatorVisible = false
                    self.activityIndicator.stopAnimating()
                    self.showSuccess("Ref: " + self.transactionIdentifier!)
                } )
                
            }else{
                UIApplication.shared.isNetworkActivityIndicatorVisible = false
                self.activityIndicator.stopAnimating()
            }
            
        })
        alert.addAction(ok)
        present(alert, animated: true, completion: nil)
    }


    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}

